designs/mfa_backup_code.rst

```diff
@@ -17,65 +17,35 @@ To implement this, we can add Backup code as a new AuthCredential, so that it ca
-
 The login session should only last for 5 mins with a Claim that allows users to change login credentials.
 
 Code Generation
 ------------------------
-The Backup code could be generated by TOTP/HOTP, or a cryptographically secure pseudorandom number generator. 
-Attackers should not be able to guess it even if they know our source code.
-
-Implementation 1
-------------------------
 
-Based on the Microsoft temporary access pass:
+The Backup code could be generated by TOTP/HOTP, or a cryptographically secure pseudorandom number generator. Attackers should not be able to guess it even if they know our source code.
 
-Add a new class/object: Backup code policy
+(See kanidmd/src/lib/utils.rs readable_password_from_random)
 
-- Each Kanidm has a Backup code policy
-- Define which users/groups can use the Backup code
-   - User can only login if they are included in the Backup code policy
-- Define parameters like max/min lifetime, one-time use and length
-- Only ``idm_hp_account_manage_priv`` can modify it?
+Implementation
+------------------------
 
 Add a new class/object: Backup code
 
-- Issue and save the Backup code in the database, it must adhere to Backup code policy
-- ``idm_account_manage_priv`` can create/read/delete it for normal users, but not themselves?
-- ``idm_hp_account_manage_priv`` can create/read/delete it for ``idm_account_manage_priv`` users and normal users?
+A user can generate one batch of Backup codes at a time, which will be saved in the database
+   - Each code in the batch can be used only once
+   - If the user generates a new batch of Backup codes, the last batch of codes become invalid.
 
-Workflow
+To invalidate the Backup code after usage, we should send an async message to the async task queue. Since the auth process is currently in a single transaction lock that is NOT writeable, using an async action allows the auth process to proceed in parallel and prevent opening nested transactions. **Note** that there is a small window between "use of the backup code" and the async action being processed to actually cause the invalidation, but this window is short enough that it's an acceptable compromise. 
+   
+Access Control
 ================
-1. User request Backup code from high-priv users (``idm_hp_account_manage_priv`` or ``idm_account_manage_priv``) via phone/email/etc.
-2. User receive Backup code from high-priv users via phone/email/etc.
-3. User login with the Backup code
 
-Implementation 2
-------------------------
-
-Add a new class/object: Backup code
-
-- Issue and save the Backup code in the database with predefined parameters of lifetime, one-time use and length
-- ``idm_hp_account_manage_priv``, ``idm_account_manage_priv`` and normal users themselves can create/read/delete it?
+We can decide which users/groups can use the Backup code via a membership to a ``backup_code_allowed`` group and making use of the existing access profile infrastructure.
```

Comment:
`idm_account_mfa_backup_code_self_priv` I think is the better name for this. For other users to admin backup codes for others it should be `idm_account_mfa_backup_code_manage_priv`. See kanidmd/src/lib/constants/entries.rs for examples of existing groups for privileges.